### ISSUE TITLE
Cell contents are slightly different for main and main2

### ISSUE THREAD

**Issue by kwalcock** [Allegra-Cohen/grid#8]
For main, these texts are in the cells shown.

  | credit / crop | season / rice | seed /science | year / high | certain /         import | people/lot
-- | -- | -- | -- | -- | -- | --
other |   | Mr. D grew   tomatoes |   |   |   |  
proportions |   |   | If a farmer had a   good price |   | Dr. Fall expects   50 |  
processes |   |   |   |   |   |  
decisions |   |   | If a farmer had a   good price |   | Dr. Fall expects   50 |  
conditions |   |   |   |   |   |  
causes |   |   | If a farmer had a   good price |   |   |  

In main2, they have moved to these positions:

  | credit / crop | season / rice | seed /science | year / high | certain /         import | people/lot
-- | -- | -- | -- | -- | -- | --
other |   |   |   | Mr. D grew   tomatoes |   |  
proportions |   |   |   | If a farmer had a   good price |   | Dr. Fall expects   50
processes |   |   |   |   |   |  
decisions |   |   |   | If a farmer had a   good price |   | Dr. Fall expects   50
conditions |   |   |   |   |   |  
causes |   |   |   | If a farmer had a   good price |   |  

I suspect the change has to do with tokenization.  main sometimes uses the tokenizer and sometimes just splits on a space.  It also retokenizes sentences that are concatenated.  main2 always uses the tokenizer and only once on each sentence.  I'm not completely certain, however.

**Comment by Allegra-Cohen** [comment]
It could also have to do with TF-IDF. In the old grid in main, I calculate TF-IDF for all the words across the entire corpus. It looks like for main2, we're calculating TF-IDF just for the anchored documents. I'm not sure which gives a better result but I have a hunch it might be better to do TF-IDF for the whole corpus.

**Comment by kwalcock** [comment]
@Allegra-Cohen, could you let me know where that is in the code?  Also, what are ps and tfps?  Are ps probabilities?  Thanks.
```python
ps_cat = calculate_ps_cat(...)
tfps = calculate_tfps(...)
```

**Comment by Allegra-Cohen** [comment]
That should be in corpus.py in the initialize function, line 32:
  `self.tfidf = self.linguist.tfidf_vectorizer.fit_transform(vector_texts)`

vector_texts is set based on corpus.documents, which are only documents returned by load_anchored_documents. In the original code, the tfidf vectorizer was applied to all the documents in the full corpus.

ps and tfps come from Mihai's 2005 paper: http://surdeanu.cs.arizona.edu/mihai/papers/kdd05.pdf. 

Yep, I used "ps" as shorthand for "probabilities." calculate_ps_cat calculates the probabilities of what the paper calls categories and what we call columns/clusters; it's calculating p(c|theta) in Eqs 8 and 9.

tfps is referring to the sums in the numerator and denominator of Eq 6, TF(...)*P(...), it's the frequency of the word in the document times whether the document is in the category in question.

Thanks!

**Comment by kwalcock** [comment]
Thanks!  So it's term frequency probabilities.

Is the original code for self.tfidf at line 70 in grid.py?

https://github.com/Allegra-Cohen/grid/blob/524e166e32af9ba48612d2847095df97416e732c/grid.py#L70

There I'm only seeing the horticulture documents.

**Comment by Allegra-Cohen** [comment]
Hi Keith,
Shoot, you're right. I was remembering an old version! Okay, so not that. Do you think the tokenizer is likely to cause such a big change? Can we check if the vectors for the moving sentences are the same? If they are, then that indicates that something in the clustering algorithm has changed, right?
Thanks!
Allegra

**Comment by kwalcock** [comment]
main2 is based on the files you first sent me all zipped up, and the outputs of the zip file and main2 are matching.  main is from after you started the repo with quite a few modifications to grid.py.  That output does not match output from the zipped files.  I've been comparing the source code and don't yet see a good reason for the discrepancy.  Do you know of anything that should have changed?  It doesn't seem like main2 is incorrect except in not having tracked the changes made between the zip file and the repo.

**Comment by Allegra-Cohen** [comment]
Hi Keith,

I see, so I must have changed something in main when I was getting rid of cluster lists, in between sending you the zip file and posting the repo. 

I've checked that the vectors match for the "Mr D. grew tomatoes" sentence in all versions, so it's got to be something in the clustering. That makes sense, because the main changes I remember was getting rid of the machine/seeded/frozen cluster lists.

If the pre-repo files and main2 match, should we chalk it up to me doing something small and dumb somewhere? I spent a long time getting the clustering algorithm in the pre-repo files right, and _much_ less time messing with it for the zip files, so if I had to guess which was right I would say main2.

**Comment by Allegra-Cohen** [comment]
I found one change, in grid_functions.py, the return statement for get_best_initial_model_k is `return ('gw', len(all_included_docs)/len_docs, allowed_clusters + modified_clusters)` in the original zip file and mathematician.py, and `return ('gw', len(all_included_docs)/len_docs, modified_clusters + allowed_clusters)` in the repo file. 

I don't think this should change the initial generation because there are no modified clusters before the user modifies them. But it might account for the frozen clusters issue I found earlier. Edit: Confirmed it doesn't change initial generation.

**Comment by kwalcock** [comment]
I noticed that as well.  Right now I'm running the zip version and the repo version side by side to try to find when they diverge.

**Comment by kwalcock** [comment]
Someone could check out the extra line at the top of cleaned_docs.csv with "sentence".  It isn't causing this problem, but maybe it adversely affects other things.

**Comment by kwalcock** [comment]
#11 should take care of this.  There was a bug added to the code between the zip and repo versions.  main and main2 should be matching again.  I will transfer some of the other changes like the `modified_clusters + allowed_clusters` in a different pass.

**Comment by Allegra-Cohen** [comment]
Ah, you found it! I'm so sorry, I totally forgot that I'd changed the label assignment code. Thank you!
